GVFS/GVFS.UnitTests/Tracing/QueuedPipeStringWriterTests.cs

```diff
@@ -0,0 +1,164 @@
+            writer.Stop();
+
+            Assert.AreEqual(1, eventCalls.Count);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[0].State);
+        }
+
+        [TestCase]
+        public void MissingPipe_RaisesStateFailing()
+        {
+            const string inputMessage = "FooBar";
+            QueuedPipeStringWriter writer = new QueuedPipeStringWriter(() => throw new Exception("Failing pipe connection"));
+
+            // Capture event invocations
+            List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs> eventCalls = new List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs>();
+            writer.StateChanged += (_, e) => eventCalls.Add(e);
+
+            // Try to write some dummy data
+            writer.Start();
+            writer.Enqueue(inputMessage);
+            writer.Stop();
+
+            Assert.AreEqual(2, eventCalls.Count);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Failing, eventCalls[0].State);
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[1].State);
+        }
+
+        [TestCase]
+        public void GoodPipe_WritesDataAndRaisesStateHealthy()
+        {
+            const string inputMessage = "FooBar";
+            byte[] expectedData = { 0x46, 0x6F, 0x6F, 0x42, 0x61, 0x72, (byte)'\n' };
+
+            string pipeName = Guid.NewGuid().ToString("N");
+
+            using (TestPipeWorker pipeWorker = new TestPipeWorker(pipeName, PipeTransmissionMode.Byte))
+            {
+                QueuedPipeStringWriter writer = new QueuedPipeStringWriter(() =>
+                    new NamedPipeClientStream(".", pipeName, PipeDirection.Out));
+
+                // Capture event invocations
+                List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs> eventCalls =
+                    new List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs>();
+                writer.StateChanged += (_, e) => eventCalls.Add(e);
+
+                // Try to write some dummy data
+                writer.Start();
+                writer.Enqueue(inputMessage);
+                writer.Stop();
+
+                byte[] actualData = pipeWorker.GetAllData();
+                CollectionAssert.AreEqual(expectedData, actualData);
+
+                Assert.AreEqual(2, eventCalls.Count);
+                Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Healthy, eventCalls[0].State);
+                Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[1].State);
+            }
+        }
+
+        private class TestPipeWorker : IDisposable
+        {
+            private readonly string pipeName;
+            private readonly PipeTransmissionMode transmissionMode;
+            private readonly Thread thread;
+            private readonly List<byte[]> allData = new List<byte[]>();
```

Comment:
Mainly a question - do we ever use the individual chunks of byte data read? Would it be less bookkeeping to just have a single list of bytes that we append to?

Reply:
Nope, we do not. But this is only test code, and it makes multiple debugging messages easier.

Reply:
> it makes multiple debugging messages easier.

Thanks - that was mainly my question - I didn't know if there was a reason for it to be implemented this way.